src/modules/status-list/status-list.service.ts

```diff
@@ -0,0 +1,204 @@
+
+@Injectable()
+export class StatusListService {
+  constructor(
+    private readonly configService: ConfigService,
+    @InjectRepository(StatusListCredential)
+    private readonly statusListRepository: Repository<StatusListCredential>
+  ) {}
+
+  /**
+   * Add StatusList2021Entry as `credentialStatus` parameter to given credential object.
+   * https://w3c-ccg.github.io/vc-status-list-2021/#statuslist2021entry
+   *
+   * @param {CredentialDto} credential
+   * @return credential with StatusList2021Entry
+   */
+  async addStatusListEntry(
+    credential: CredentialDto
+  ): Promise<CredentialWithStatusDto> {
+    if (await this.isCredentialRevoked(credential.id)) {
+      throw new ConflictException('Credential already revoked');
+    }
+
+    const statusListCredential = new URL(
+      `${STATUS_LIST_MODULE_PATH}/${credential.id}`,
+      this.configService.get('STATUS_LIST_DOMAIN')
+    );
+
+    const newStatusList = StatusListCredential.create({
+      credentialId: credential.id,
+      namespace: credential.credentialSubject.role.namespace,
+    });
+
+    await this.statusListRepository.save(newStatusList);
+
+    // https://w3c-ccg.github.io/vc-status-list-2021/#statuslist2021entrya
+    const credentialStatus: StatusList2021EntryDto = {
+      type: 'StatusList2021Entry',
+      statusPurpose: 'revocation',
+      statusListIndex: '0', // because we are using one-to-one mapping
+      statusListCredential: statusListCredential.href,
+    };
+
+    return { ...credential, credentialStatus };
+  }
+
+  /**
+   * Create a new StatusList2021Credential object for rejection purposes of a given credential for revoker.
+   * Object is not signed and not saved to database.
+   * https://w3c-ccg.github.io/vc-status-list-2021/#statuslist2021credential
+   *
+   * @param {VerifiableCredentialDto} credential
+   * @param {String} currentUser
+   * @return StatusList2021Credential without proof
+   */
+  async createStatusListCredential(
+    credential: VerifiableCredentialDto,
+    currentUser: string
+  ): Promise<StatusListCredentialDto> {
+    const statusList = await this.fetchStatusList(credential.id);
+
+    if (!statusList) {
+      throw new BadRequestException('Credential was not registered');
+    }
+
+    if (statusList.credentialStatus) {
+      throw new ConflictException('Credential already revoked');
+    }
+
+    if (statusList.namespace !== credential.credentialSubject.role.namespace) {
+      throw new BadRequestException(
+        'Credential was registered with another namespace'
+      );
+    }
+
+    const array = Uint8Array.from([1]);
+    const encodedList = Buffer.from(gzip(array)).toString('base64');
+
+    // https://w3c-ccg.github.io/vc-status-list-2021/#statuslist2021credential
+    const statusListCredential = {
+      '@context': [
+        'https://www.w3.org/2018/credentials/v1',
+        'https://w3id.org/vc/status-list/2021/v1',
+      ],
+      id: credential.credentialStatus.statusListCredential,
```

Comment:
This seems confusing and maybe wrong to me. An issue that I am imagining that that what if the passed in credential had a `credential.id` that we had registered, the `namespace` was correct, the proof verified, but somehow the `credential.credentialStatus.statusListCredential` was wrong. I'm not sure it's really a security issue, just an opportunity to get weird behaviour.

Maybe we instead we should:
1. use the whole URL as our `credential.id` in `addStatusListEntry`? Then we can just do `statusList.id` here.
2. Or maybe we should extract a function that creates the URL from the id and configuration and then we can call it here. The downside to that might be that if the configuration values ever change, it would break the existing statusLists
3. Or maybe we store the components of the URL in the `CredentialStatusList` entity, then we can reconstruct the `statusListCredential.id` from our persisted values while also being able to run a query like `SELECT * WHERE STATUS_LIST_DOMAIN='https://identity-cache-dev.energyweb.org'`.

I guess I'm leaning towards option 3 at the moment. What are your thoughts @Harasz ?

Reply:
I agree, options 3